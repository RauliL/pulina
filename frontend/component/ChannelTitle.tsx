import * as React from 'react';
import { Button, Navbar } from 'reactstrap';

import { Channel } from '../types/channel';

export interface Props {
  channel: Channel;
  onToggleChannelList: () => void;
  onToggleUserList: () => void;
}

const ChannelTitle: React.SFC<Props> = (props) => (
  <Navbar className="border-bottom">
    <span className="navbar-brand">{props.channel.name}</span>
    <div className="float-right d-lg-none">
      <Button
        outline={true}
        className="mr-2"
        onClick={props.onToggleChannelList}
      >
        <i className="fas fa-hashtag"/>
      </Button>
      <Button outline={true} onClick={props.onToggleUserList}>
        <i className="fas fa-user"/>
      </Button>
    </div>
  </Navbar>
);

export default ChannelTitle;
