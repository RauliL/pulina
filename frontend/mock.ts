import { Channel } from './types/channel';

export const mockChannel: Channel = {
  name: '#test-channel',
  users: [],
  log: [],
  hasUnreadMessages: false,
  hasUnreadHighlight: false,
};
