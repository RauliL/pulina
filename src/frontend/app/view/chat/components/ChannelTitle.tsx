import React, { FunctionComponent } from "react";
import { useDispatch } from "react-redux";
import { Button, Navbar } from "reactstrap";

import { Channel } from "../../../../client";
import { AppDispatch, uiSlice } from "../../../../store";

export type ChannelTitleProps = {
  channel: Channel;
};

const ChannelTitle: FunctionComponent<ChannelTitleProps> = ({ channel }) => {
  const dispatch = useDispatch<AppDispatch>();

  const handleChannelListToggleClick = () => {
    dispatch(uiSlice.actions.toggleChannelList());
  };

  const handleUserListToggleClick = () => {
    dispatch(uiSlice.actions.toggleUserList());
  };

  return (
    <Navbar className="border-bottom">
      <span className="navbar-brand">{channel.name}</span>
      {channel.topic && <span>{channel.topic}</span>}
      <div className="float-right d-lg-none">
        <Button
          outline
          className="btn-toggle-channel-list mr-2"
          onClick={handleChannelListToggleClick}
        >
          <i className="fas fa-hashtag" />
        </Button>
        <Button
          outline
          className="btn-toggle-user-list"
          onClick={handleUserListToggleClick}
        >
          <i className="fas fa-user" />
        </Button>
      </div>
    </Navbar>
  );
};

ChannelTitle.displayName = "ChannelTitle";

export default ChannelTitle;
