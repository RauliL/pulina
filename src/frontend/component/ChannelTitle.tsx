import React, { FunctionComponent } from "react";
import { Button, Navbar } from "reactstrap";

import { Channel } from "../types/channel";

export type Props = {
  channel: Channel;
  onToggleChannelList: () => void;
  onToggleUserList: () => void;
};

const ChannelTitle: FunctionComponent<Props> = (props) => (
  <Navbar className="border-bottom">
    <span className="navbar-brand">{props.channel.name}</span>
    <div className="float-right d-lg-none">
      <Button
        outline={true}
        className="btn-toggle-channel-list mr-2"
        onClick={props.onToggleChannelList}
      >
        <i className="fas fa-hashtag" />
      </Button>
      <Button
        outline={true}
        className="btn-toggle-user-list"
        onClick={props.onToggleUserList}
      >
        <i className="fas fa-user" />
      </Button>
    </div>
  </Navbar>
);

export default ChannelTitle;
