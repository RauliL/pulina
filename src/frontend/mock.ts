import { ClientEvent, ClientEventType, ListResult } from "../common";
import {
  Channel,
  ChannelUser,
  Client,
  LogEntry,
  LogEntryType,
} from "./client";
import { ClientState, RootState, UIState } from "./store";

export const mockChannelUser: Readonly<ChannelUser> = {
  isOperator: true,
  nick: "test-user",
};

export const mockChannel: Readonly<Channel> = {
  name: "#test-channel",
  hasUnreadHighlights: false,
  hasUnreadMessages: false,
  log: [],
  topic: null,
  users: [mockChannelUser],
};

export const mockListResult: Readonly<ListResult> = {
  channel: "#test-channel",
  population: 1,
  topic: null,
};

export const mockLogEntry: Readonly<LogEntry> = {
  id: "2ba52fc8-bf86-11ef-a759-23984454453d",
  isHighlight: false,
  source: "test-user",
  text: "test text",
  timestamp: 1734776864844,
  type: LogEntryType.MESSAGE,
};

export type MockClient = Client & {
  events: ClientEvent[];
  reset: () => void;
};

export const mockClient = (): MockClient => {
  const events: ClientEvent[] = [];

  return {
    events,

    list(query?: string) {
      this.send({
        type: ClientEventType.LIST,
        query: query ?? null,
      });
    },

    reset() {
      events.length = 0;
    },

    send(event) {
      events.push(event);
    },
  };
};

export const mockClientState: Readonly<ClientState> = {
  channels: {},
};

export const mockUIState: Readonly<UIState> = {
  showChannelList: false,
  showUserList: false,
};

export const mockRootState: Readonly<RootState> = {
  client: mockClientState,
  ui: mockUIState,
};
