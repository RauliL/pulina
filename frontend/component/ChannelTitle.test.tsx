import { mount, shallow } from 'enzyme';
import { defaultTo, noop } from 'lodash';
import * as React from 'react';
import { Button, Navbar } from 'reactstrap';

import { mockChannel } from '../mock';

import ChannelTitle, { Props } from './ChannelTitle';

describe('<ChannelTitle/>', () => {
  const renderComponent = (props: Partial<Props> = {}) => (
    <ChannelTitle
      channel={defaultTo(props.channel, mockChannel)}
      onToggleChannelList={defaultTo(props.onToggleChannelList, noop)}
      onToggleUserList={defaultTo(props.onToggleUserList, noop)}
    />
  );
  const mountComponent = (props: Partial<Props> = {}) => mount(
    renderComponent(props),
  );

  it('should render without throwing exception', () => {
    expect(shallow(renderComponent())).toMatchSnapshot();
  });

  it('should render <Navbar/>', () => {
    expect(mountComponent().find(Navbar)).toHaveLength(1);
  });

  it('should render the channel name', () => {
    const component = mountComponent();

    expect(component.find('.navbar-brand').text()).toBe(mockChannel.name);
  });

  it('should render two <Button/>s', () => {
    expect(mountComponent().find(Button)).toHaveLength(2);
  });

  it('should invoke callback when channel list button is clicked', () => {
    const onToggleChannelList = jest.fn();
    const component = mountComponent({ onToggleChannelList });

    component.find('.btn-toggle-channel-list').hostNodes().simulate('click');
    expect(onToggleChannelList).toBeCalled();
  });

  it('should invoke callback when user list button is clicked', () => {
    const onToggleUserList = jest.fn();
    const component = mountComponent({ onToggleUserList });

    component.find('.btn-toggle-user-list').hostNodes().simulate('click');
    expect(onToggleUserList).toBeCalled();
  });
});
