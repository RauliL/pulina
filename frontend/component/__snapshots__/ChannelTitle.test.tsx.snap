// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChannelTitle/> should render without throwing exception 1`] = `
<Navbar
  className="border-bottom"
  expand={false}
  tag="nav"
>
  <span
    className="navbar-brand"
  >
    #test-channel
  </span>
  <div
    className="float-right d-lg-none"
  >
    <Button
      className="btn-toggle-channel-list mr-2"
      color="secondary"
      onClick={[Function]}
      outline={true}
      tag="button"
    >
      <i
        className="fas fa-hashtag"
      />
    </Button>
    <Button
      className="btn-toggle-user-list"
      color="secondary"
      onClick={[Function]}
      outline={true}
      tag="button"
    >
      <i
        className="fas fa-user"
      />
    </Button>
  </div>
</Navbar>
`;
